include/mxnet/op_attr_types.h

```diff
@@ -208,6 +208,17 @@ using FCreateOpState = std::function<OpStatePtr (const NodeAttrs& attrs,
                                                  Context ctx,
                                                  const mxnet::ShapeVector& in_shape,
                                                  const std::vector<int>& in_type)>;
+
+/*!
+ * \brief Whether the operator always produces the same
+ *        output given the same input.
+ *        This enables certain optimizations
+ *        like common expression elimination.
+ *
+ * \note Register under "FHasDeterministicOutput"
+ */
+using FHasDeterministicOutput = bool;
```

Comment:
This is a per-op attribute, right? Are we supposed to mark relevant operators with the right value for this attribute in this PR?

Reply:
This is an option for potential optimization - if the op uses state or requests resources then it is not considered for this optimization. This option enables marking such op as safe for deduplicating (an example of this would be Convolution, which asks for resources, but it is only a workspace). I will add this option to such relevant operators.